File: analytics/src/main/java/io/appmetrica/analytics/impl/crash/client/converter/NativeCrashConverter.java
package io.appmetrica.analytics.impl.crash.client.converter;

import androidx.annotation.NonNull;
import io.appmetrica.analytics.coreapi.internal.data.ProtobufConverter;
import io.appmetrica.analytics.coreutils.internal.WrapUtils;
import io.appmetrica.analytics.impl.crash.ndk.NativeCrashHandlerDescription;
import io.appmetrica.analytics.impl.crash.ndk.NativeCrashModel;
import io.appmetrica.analytics.impl.crash.ndk.NativeCrashSource;
import io.appmetrica.analytics.impl.protobuf.backend.CrashAndroid;
import java.util.EnumMap;

public class NativeCrashConverter implements ProtobufConverter<NativeCrashModel, CrashAndroid.Crash> {

    private static final EnumMap<NativeCrashSource, Integer> handlerMapping;

    static {
        handlerMapping = new EnumMap<>(NativeCrashSource.class);
        handlerMapping.put(NativeCrashSource.UNKNOWN, CrashAndroid.UNKNOWN);
        handlerMapping.put(NativeCrashSource.CRASHPAD, CrashAndroid.CRASHPAD);
    }

    @NonNull
    @Override
    public CrashAndroid.Crash fromModel(@NonNull NativeCrashModel value) {
        CrashAndroid.Crash crash = new CrashAndroid.Crash();
        crash.type = CrashAndroid.Crash.NATIVE;
        crash.native_ = new CrashAndroid.Crash.NativeCrash();
        crash.native_.nativeCrashPayload = value.getData();
        NativeCrashHandlerDescription description = value.getHandlerDescription();
        crash.native_.handler = new CrashAndroid.NativeCrashHandler();
        Integer library = handlerMapping.get(description.getSource());
        if (library != null) {
            crash.native_.handler.source = library;
        }
        crash.native_.handler.version = WrapUtils.getOrDefault(description.getHandlerVersion(), "");
        return crash;
    }

    @NonNull
    @Override
    public NativeCrashModel toModel(@NonNull CrashAndroid.Crash nano) {
        throw new UnsupportedOperationException();
    }
}

File: analytics/src/main/java/io/appmetrica/analytics/impl/crash/jvm/converter/NativeCrashConverter.kt
package io.appmetrica.analytics.impl.crash.jvm.converter

import io.appmetrica.analytics.coreapi.internal.data.ProtobufConverter
import io.appmetrica.analytics.impl.crash.ndk.NativeCrashModel
import io.appmetrica.analytics.impl.protobuf.backend.CrashAndroid
import io.appmetrica.analytics.impl.protobuf.backend.CrashAndroid.Crash
import io.appmetrica.analytics.ndkcrashesapi.internal.NativeCrashSource

internal class NativeCrashConverter : ProtobufConverter<NativeCrashModel, Crash> {
    companion object {
        private val handlerMapping = mapOf(
            NativeCrashSource.UNKNOWN to CrashAndroid.UNKNOWN,
            NativeCrashSource.CRASHPAD to CrashAndroid.CRASHPAD,
        )
    }

    override fun fromModel(value: NativeCrashModel): Crash {
        return Crash().apply {
            type = Crash.NATIVE
            native_ = Crash.NativeCrash().apply {
                nativeCrashPayload = value.data
                handler = CrashAndroid.NativeCrashHandler().apply {
                    handlerMapping[value.handlerDescription.source]?.let { source = it }
                    version = value.handlerDescription.handlerVersion ?: ""
                }
            }
        }
    }

    override fun toModel(value: Crash): NativeCrashModel {
        throw UnsupportedOperationException()
    }
}
